HLTrigger/JetMET/plugins/HLTHFJetCleaner.cc

```diff
@@ -0,0 +1,268 @@
+// -*- C++ -*-
+//
+// Package:    HLTrigger/HLTHFJetCleaner
+// Class:      HLTHFJetCleaner
+//
+/**\class HLTHFJetCleaner HLTHFJetCleaner.cc HLTrigger/HLTHFJetCleaner/plugins/HLTHFJetCleaner.cc
+
+ Description: [one line class summary]
+
+ Implementation:
+     [Notes on implementation]
+*/
+//
+// Original Author:  Alp Akpinar
+//         Created:  Wed, 01 Dec 2021 15:47:19 GMT
+//
+//
+
+// system include files
+#include <memory>
+
+// user include files
+#include "FWCore/Framework/interface/Frameworkfwd.h"
+#include "FWCore/Framework/interface/stream/EDProducer.h"
+
+#include "FWCore/Framework/interface/Event.h"
+#include "FWCore/Framework/interface/MakerMacros.h"
+
+#include "FWCore/ParameterSet/interface/ParameterSet.h"
+#include "FWCore/Utilities/interface/StreamID.h"
+
+#include "DataFormats/JetReco/interface/PFJet.h"
+#include "DataFormats/METReco/interface/CaloMET.h"
+#include "DataFormats/Math/interface/deltaPhi.h"
+
+//
+// class declaration
+//
+
+class HLTHFJetCleaner : public edm::stream::EDProducer<> {
+public:
+  typedef std::vector<reco::PFJet> JetCollection;
```

Comment:
Can this plugin make sense for non-PF jets (e.g. Calo) ? If so, it should be templated. If templated, all should go under `plugins/*.h`, and the instantiations in `plugins/SealModule.cc`.

Reply:
Thanks for the comments, this plugin is meant to be run on PF jets that are reconstructed during the HLT path sequence, that's why I specified `reco::PFJet` in the code. But the methods accessed for jets are implemented in the base class, `reco::Jet`, so for better generalization, would it work if I replace the `reco::PFJet` objects with the base `reco::Jet` object?

Reply:
This should work, or at least, I would try it.

(if there is an expectation that this plugin will evolve in a way that reco::Jet wont suffice anymore, for example if one will need the option to access class members only available for PFJets, then it should be templated; but if this is not expected to be needed for now, reco::Jet would be the simplest option)

Reply:
Hi @missirol, while running locally, I've encountered an issue if I use the `reco::Jet` class. I believe the issue is that the module tries to declare a vector of `reco::Jet` objects. Here is the error message:

```
Exception Message:
No data dictionary found for the following classes:

  edm::Wrapper<std::vector<reco::Jet> >
  std::vector<reco::Jet>
```

This problem does not happen with `reco::PFJet`. Do you know if there is a workaround for this? If not, would it be possible to stick with `reco::PFJet`, given that we intend to run the plugin on those objects. Thanks!

Reply:
I think the bottom line (which I partly overlooked in one of my comments yesterday) is that we likely want to produce as output the same type of collection which is used in input (e.g. `reco::PFJet`).

> If not, would it be possible to stick with reco::PFJet, given that we intend to run the plugin on those objects.

Sure, but in that case I would prefer to also support `reco::CaloJet` as usually done for similar plugins. I suggest to do the following.

 - Template the class with 1 template argument (the "jet_type") used to define the type of input and output jets. The whole template will live in one header file, named `plugins/HLTJetHFCleaner.h`.

 - Add 2 instantiations of this template in `plugins/SealModule.cc`, similarly to [this](https://github.com/cms-sw/cmssw/blob/f6c204bfb2f92e827bac326385786fed876046e0/HLTrigger/JetMET/plugins/SealModule.cc#L252), yielding `HLTCaloJetHFCleaner` and `HLTPFJetHFCleaner`.

Reply:
I understand, thanks I can do that! So I'll include the templated class under `plugins/HLTJetHFCleaner.h`, and add 2 
2 instantiations under `plugins/SealModule.cc`. I can do something similar to:
```cpp
typedef HLTCaloJetHFCleaner HLTJetHFCleaner<reco::CaloJet>;
DEFINE_FWK_MODULE(HLTCaloJetHFCleaner);

typedef HLTPFJetHFCleaner HLTJetHFCleaner<reco::PFJet>;
DEFINE_FWK_MODULE(HLTPFJetHFCleaner);
```

Reply:
Thanks for the suggestion, this is implemented in [5cb552b](https://github.com/cms-sw/cmssw/pull/37539/commits/5cb552b77e666f204527ed68910d6b9a4ead4119). Please let me know if you'd like any changes.